### ISSUE TITLE
bigquery/storage/managedwriter: support schema change notifications

### ISSUE THREAD

**Issue by shollyman** [googleapis/google-cloud-go#5244]
`cloud.google.com/go/bigquery/storage/managedwriter` doesn't yet include support for dealing with schema change notifications from the storage write backend.

When schema changes are propagated from the v1 storage write backend (e.g. from a `tables.update` or `ALTER TABLE` query), the backend service populates `UpdatedSchema` as part of the [`AppendRowsResponse`](https://pkg.go.dev/google.golang.org/genproto/googleapis/cloud/bigquery/storage/v1#AppendRowsResponse), which is part of the core AppendRows bidi stream.  There's no other notification path, barring doing something horrible like calling the unary `GetWriteStream` and polling for changes.

The managedwriter veneer package already allows users to set a new proto descriptor (schema) when appending by passing the [`UpdateSchemaDescriptor`](https://pkg.go.dev/cloud.google.com/go/bigquery/storage/managedwriter#UpdateSchemaDescriptor) option to the AppendRows method on a ManagedStream instance.

The question to tackle here is *how the notifications should be surfaced to users*.

The **least effort** solution is to add the updated schema to the [`AppendResult`](https://pkg.go.dev/cloud.google.com/go/bigquery/storage/managedwriter#AppendResult) future, and simply return the schema as a method there.  We'd use the same channel block as we use for Ready/GetResult, so to be performant users would need to go through extra hoops to check every response

However, another option is to define an interface for propagating the notifications, and allowing users to register a listener capable of receiving the callback.  This obviates the need for users to monitor responses for the notification directly, but it's unclear what should be passed back.

### defining the contract 

Clearly, the **new schema** needs to be part of the message.  This is yet another instance of where the v1 proto types need to either be exposed directly from managedwriter, or we need to wrap; I'm inclined to give up on encapsulating the v1 proto messages and simply couple managedwriter to the v1 API directly.

Another component of the notification likely should be the **stream name** itself.  The backend does not currently support stream multiplexing, but it is expected to in the future.  To disambiguate between write streams, we should callback with the identifier.

I not sure **context** should be part of the notification at this point.  The context we'd have available for use is the retained context inside ManagedStream responsible for network stream management, and it seems odd to use that to notify the likely owner/controller of the ManagedStream instance.  Then again, if we want to end up propagating things like trace spans in the future, it's probably necessary.  We could generate a new context, or derive one from the retained context if we end up sending one.

The interesting bit is whether the notification should contain a reference to the instance of ManagedStream.  My current feeling is that this should **not** be part of the notification.  Users can't really do anything useful with the new schema unless they have data to send, so a reference to the writer is already necessarily being retained elsewhere.

### other considerations

How/when should callback registration be supported?  If we support a single registration, it makes sense that is defined when the ManagedStream is constructed, and added as a WriterOption.  If we allow multiple registrations, we probably want the ability to add/remove callbacks on the `ManagedStream`, or via a new `AppendOption` when user calls `AppendRows()`.

Should the callback return an error?  What does error presence do to the `ManagedStream`?  Does it mean we should retry the callback?  Does it mean we should unregister the receiver for future changes?

Blocking behavior:  the nature of the existing bidi contract makes it infeasible to block new appends while sending the schema notification (we could have already sent multiple appends before the response arrives), so this would not be in scope.

Schema compatibility:  incompatible schema is out of scope here; if you've managed to create an incompatible schema you're likely doing things like replacing a table, which means the existing stream IDs are already defunct.  default streams may be a special case, but we're concerned with compatible modifications here.

### tl;dr scribbles

```
type SchemaListener interface {
  SchemaUpdated(ctx context.Context, streamName string, newSchema storagepb.TableSchema) error
}

func WithSchemaNotification(listener SchemaListener) WriterOption {
 ...
}
```

**Comment by shollyman** [comment]
Related: internal issue 211899346 filed related to UpdatedSchema behaviors in practice.

**Comment by shollyman** [comment]
Absent strong feedback on this, I'm closing for now but we can revisit it in the face of further updates.
